src/coprocessor/dag/batch_executor/table_scan_executor.rs

```diff
@@ -0,0 +1,906 @@
+// Copyright 2019 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::sync::Arc;
+
+use cop_datatype::{EvalType, FieldTypeAccessor};
+use kvproto::coprocessor::KeyRange;
+use tipb::expression::FieldType;
+use tipb::schema::ColumnInfo;
+
+use crate::storage::Store;
+use crate::util::collections::HashMap;
+
+use super::interface::*;
+use crate::coprocessor::codec::batch::{LazyBatchColumn, LazyBatchColumnVec};
+use crate::coprocessor::dag::expr::{EvalConfig, EvalContext};
+use crate::coprocessor::dag::Scanner;
+use crate::coprocessor::Result;
+
+pub struct BatchTableScanExecutor<S: Store>(
+    super::scan_executor::ScanExecutor<
+        S,
+        TableScanExecutorImpl,
+        super::ranges_iter::PointRangeEnable,
+    >,
+);
+
+impl<S: Store> BatchTableScanExecutor<S> {
+    pub fn new(
+        store: S,
+        config: Arc<EvalConfig>,
+        columns_info: Vec<ColumnInfo>,
+        key_ranges: Vec<KeyRange>,
+        desc: bool,
+    ) -> Result<Self> {
+        let is_column_filled = vec![false; columns_info.len()];
+        let mut key_only = true;
+        let mut handle_index = None;
+        let mut schema = Vec::with_capacity(columns_info.len());
+        let mut columns_default_value = Vec::with_capacity(columns_info.len());
+        let mut column_id_index = HashMap::default();
+
+        for (index, mut ci) in columns_info.into_iter().enumerate() {
+            schema.push(super::scan_executor::field_type_from_column_info(&ci));
+            columns_default_value.push(ci.take_default_val());
+
+            if ci.get_pk_handle() {
+                handle_index = Some(index);
+            } else {
+                key_only = false;
+                column_id_index.insert(ci.get_column_id(), index);
+            }
+
+            // Note: if two PK handles are given, we will only preserve the *last* one. Also if two
+            // columns with the same column id are given, we will only preserve the *last* one.
+        }
+
+        let imp = TableScanExecutorImpl {
+            context: EvalContext::new(config),
+            schema,
+            columns_default_value,
+            column_id_index,
+            key_only,
+            handle_index,
+            is_column_filled,
```

Comment:
Please set it directly so we needn't transfer the value twice(first on L46)

Reply:
The information it needs (`columns_info`) is consumed so it must be constructed beforehand. Additionally I don't think there could be performance differences even if it can be placed here.

Reply:
It was mentioned in the sharing meeting by our CEO yesterday, did you miss it?

Reply:
The performance characteristics of Go and Rust are quite different due to no-GC and LLVM, techniques that work in Go don't necessarily make sense in Rust. 

The `is_column_filled` value is a `Vec<bool>` which is non-copyable. This is *moved*, meaning *only 3 pointers will be copied* and nothing else will happen.

Reply:
@AndreMouche I think LLVM is clever enough to figure out that the variable constructed previously is used directly as a field so that the allocation address will be just inside the field. It's very simple compiling optimizations. Even if the compiler is not clever enough to figure this out, nothing cost will actually happen in Rust, as explained by @kennytm above.

Reply:
OK